### PR TITLE
Ci full tests

### PR DESCRIPTION
**Pull request by adrianre12**
### Description:
I have been working on getting CI to run all the tests. To do this it means enabling virtual rendering for each platform. I have had variable success with this and could do with some more brains on it.
**Windows**

I Installed [OSmesa ](https://github.com/pal1000/mesa-dist-win )and the tests now all pass. 

The default install script needs some work, but that is in progress upstream.

It is a bit slow but I am working on that.

**MacOS**

 I don't have a clue about MacOS. I can't work out how to even diagnose the problem.

`Unable to create basic Accelerated OpenGL renderer.
Unable to create basic Accelerated OpenGL renderer.
Core Image is now using the software OpenGL renderer. This will be slow.
GLFW: An invalid error was not accepted by the caller: FormatUnavailable: NSGL: Failed to find a suitable pixel format
GLFW: Please report this bug in the Go package immediately.
panic: FormatUnavailable: NSGL: Failed to find a suitable pixel format`

**Linux**

This should have been the easiest, just run the tests with Xvfb. It does work well, except for random failures in the glfw tests. They seem to cluster around cursor and tappable tests. It looks like a race condition between render and test, but I am not sure. Running the same Xvfb setup on my Linux Mint VM runs fine except if I can put the host under high load, then I see an occasional error.

### Checklist:
- [X ] Tests included.
- [X ] Lint and formatter run with no errors.
- [ ] Tests all pass.

### CODE REVIEW

**Comment by adrianre12** [comment]
I have build a Ubuntu 20.04 server VM, so far it seems much easier to get it to fail.

**Comment by adrianre12** [comment]
I have installed the libosmesa6-dev pkg on my server but it does not make any difference, I assume this is because GLFW is not looking for it. In the 3.3 release notes it mentions it but I don't understand how to set it up. The release notes say:
`OSMesa off-screen context creation support
GLFW now supports creating off-screen OpenGL contexts using OSMesa by setting GLFW_CONTEXT_CREATION_API to GLFW_OSMESA_CONTEXT_API. Native access function have been added to retrieve the OSMesa color and depth buffers.

There is also a new null backend that uses OSMesa as its native context creation API, intended for automated testing. This backend does not provide input.`

**Comment by AlbinoGeek** [comment]
@adrianre12 OSMesa appears to be a Linux-only thing atm:
https://www.glfw.org/docs/3.3/compile.html#compile_deps_osmesa

Which would use the compilation option:
- `_GLFW_OSMESA` to use the OSMesa API (headless and non-interactive)

**I don't think you want this.** `fyne` is not `headless`, and definitely requires `interactivity`. I don't think we support or utilize off-screen rendering yet, but only a core team member who works on the drivers would be able to answer that for sure.

**Comment by adrianre12** [comment]
I modified TestWindow_Cursor so that it logged more info
`func TestWindow_Cursor(t *testing.T) {
	w := createWindow("Test").(*window)
	e := widget.NewEntry()
	u, _ := url.Parse("https://testing.fyne")
	h := widget.NewHyperlink("Testing", u)
	b := widget.NewButton("Test", nil)

	w.SetContent(container.NewVBox(e, h, b))

	w.mouseMoved(w.viewport, 10, float64(e.Position().Y+10))
	textCursor := desktop.TextCursor
	t.Logf("Content=%+v", w.Canvas().Content())
	t.Logf("w=%+v", w)

	if w.cursor != textCursor {
		t.Logf("cursor wrong %+v, sleeping", w.cursor)
		time.Sleep(1000 * time.Millisecond)
		t.Logf("cursor=%+v", w.cursor)
		t.Logf("Content=%+v", w.Canvas().Content())
		t.Logf("w=%+v", w)
	}

	assert.Equal(t, textCursor, w.cursor)`

Running the test multiple times I got
`=== RUN   TestWindow_Cursor
    window_test.go:222: Content=&{size:{Width:61 Height:119} position:{X:4 Y:4} Hidden:false Layout:0xc0000c7848 Objects:[0xc0000c9500 0xc000144580 0xc00009ad20]}
    window_test.go:223: w=&{viewport:0xc00030a630 viewLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} createLock:{done:0 m:{state:0 sema:0}} decorate:true fixedSize:false cursor:1 customCursor:<nil> canvas:0xc000298700 title:Test icon:<nil> mainmenu:<nil> clipboard:<nil> master:false fullScreen:false centered:false visible:false mousePos:{X:10 Y:10} mouseDragged:<nil> mouseDraggedObjStart:{X:0 Y:0} mouseDraggedOffset:{X:0 Y:0} mouseDragPos:{X:0 Y:0} mouseDragStarted:false mouseButton:0 mouseOver:<nil> mouseLastClick:<nil> mousePressed:<nil> mouseClickCount:0 mouseCancelFunc:<nil> onClosed:<nil> onCloseIntercepted:<nil> menuTogglePending: menuDeactivationPending: xpos:0 ypos:0 width:69 height:127 shouldExpand:true eventLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} eventQueue:0xc0001007e0 eventWait:{noCopy:{} state1:[0 0 0]} pending:[]}
--- PASS: TestWindow_Cursor (0.05s)
=== RUN   TestWindow_Cursor
    window_test.go:222: Content=&{size:{Width:1 Height:1} position:{X:4 Y:4} Hidden:false Layout:0xc000180630 Objects:[0xc000322c00 0xc00033a280 0xc00033c140]}
    window_test.go:223: w=&{viewport:0xc000241320 viewLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} createLock:{done:0 m:{state:0 sema:0}} decorate:true fixedSize:false cursor:0 customCursor:<nil> canvas:0xc000298800 title:Test icon:<nil> mainmenu:<nil> clipboard:<nil> master:false fullScreen:false centered:false visible:false mousePos:{X:10 Y:10} mouseDragged:<nil> mouseDraggedObjStart:{X:0 Y:0} mouseDraggedOffset:{X:0 Y:0} mouseDragPos:{X:0 Y:0} mouseDragStarted:false mouseButton:0 mouseOver:<nil> mouseLastClick:<nil> mousePressed:<nil> mouseClickCount:0 mouseCancelFunc:<nil> onClosed:<nil> onCloseIntercepted:<nil> menuTogglePending: menuDeactivationPending: xpos:0 ypos:0 width:69 height:127 shouldExpand:true eventLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} eventQueue:0xc000100fc0 eventWait:{noCopy:{} state1:[0 0 0]} pending:[]}
    window_test.go:226: cursor wrong 0, sleeping
    window_test.go:228: cursor=0
    window_test.go:229: Content=&{size:{Width:61 Height:119} position:{X:4 Y:4} Hidden:false Layout:0xc000180630 Objects:[0xc000322c00 0xc00033a280 0xc00033c140]}
    window_test.go:230: w=&{viewport:0xc000241320 viewLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} createLock:{done:0 m:{state:0 sema:0}} decorate:true fixedSize:false cursor:0 customCursor:<nil> canvas:0xc000298800 title:Test icon:<nil> mainmenu:<nil> clipboard:<nil> master:false fullScreen:false centered:false visible:false mousePos:{X:10 Y:10} mouseDragged:<nil> mouseDraggedObjStart:{X:0 Y:0} mouseDraggedOffset:{X:0 Y:0} mouseDragPos:{X:0 Y:0} mouseDragStarted:false mouseButton:0 mouseOver:<nil> mouseLastClick:<nil> mousePressed:<nil> mouseClickCount:0 mouseCancelFunc:<nil> onClosed:<nil> onCloseIntercepted:<nil> menuTogglePending: menuDeactivationPending: xpos:0 ypos:0 width:69 height:127 shouldExpand:false eventLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} eventQueue:0xc000100fc0 eventWait:{noCopy:{} state1:[0 0 0]} pending:[]}
    window_test.go:233:
                Error Trace:    window_test.go:233
                Error:          Not equal:
                                expected: 1
                                actual  : 0
                Test:           TestWindow_Cursor
--- FAIL: TestWindow_Cursor (1.05s)
=== RUN   TestWindow_Cursor
    window_test.go:222: Content=&{size:{Width:1 Height:1} position:{X:4 Y:4} Hidden:false Layout:0xc000181208 Objects:[0xc000322d80 0xc00033a880 0xc00033c1e0]}
    window_test.go:223: w=&{viewport:0xc00030a900 viewLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} createLock:{done:0 m:{state:0 sema:0}} decorate:true fixedSize:false cursor:0 customCursor:<nil> canvas:0xc000298900 title:Test icon:<nil> mainmenu:<nil> clipboard:<nil> master:false fullScreen:false centered:false visible:false mousePos:{X:10 Y:10} mouseDragged:<nil> mouseDraggedObjStart:{X:0 Y:0} mouseDraggedOffset:{X:0 Y:0} mouseDragPos:{X:0 Y:0} mouseDragStarted:false mouseButton:0 mouseOver:<nil> mouseLastClick:<nil> mousePressed:<nil> mouseClickCount:0 mouseCancelFunc:<nil> onClosed:<nil> onCloseIntercepted:<nil> menuTogglePending: menuDeactivationPending: xpos:0 ypos:0 width:69 height:127 shouldExpand:true eventLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} eventQueue:0xc0001011a0 eventWait:{noCopy:{} state1:[0 0 0]} pending:[]}
    window_test.go:226: cursor wrong 0, sleeping
    window_test.go:228: cursor=0
    window_test.go:229: Content=&{size:{Width:61 Height:119} position:{X:4 Y:4} Hidden:false Layout:0xc000181208 Objects:[0xc000322d80 0xc00033a880 0xc00033c1e0]}
    window_test.go:230: w=&{viewport:0xc00030a900 viewLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} createLock:{done:0 m:{state:0 sema:0}} decorate:true fixedSize:false cursor:0 customCursor:<nil> canvas:0xc000298900 title:Test icon:<nil> mainmenu:<nil> clipboard:<nil> master:false fullScreen:false centered:false visible:false mousePos:{X:10 Y:10} mouseDragged:<nil> mouseDraggedObjStart:{X:0 Y:0} mouseDraggedOffset:{X:0 Y:0} mouseDragPos:{X:0 Y:0} mouseDragStarted:false mouseButton:0 mouseOver:<nil> mouseLastClick:<nil> mousePressed:<nil> mouseClickCount:0 mouseCancelFunc:<nil> onClosed:<nil> onCloseIntercepted:<nil> menuTogglePending: menuDeactivationPending: xpos:0 ypos:0 width:69 height:127 shouldExpand:false eventLock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} eventQueue:0xc0001011a0 eventWait:{noCopy:{} state1:[0 0 0]} pending:[]}
    window_test.go:233:
                Error Trace:    window_test.go:233
                Error:          Not equal:
                                expected: 1
                                actual  : 0
                Test:           TestWindow_Cursor
--- FAIL: TestWindow_Cursor (1.07s)
`
This shows when the test fails the content size has not been set correctly `size:{Width:1 Height:1}` but 1s later it has been set correctly {Width:61 Height:119}

**Comment by adrianre12** [comment]
I modified TestWindow_Cursor to
`func TestWindow_Cursor(t *testing.T) {
	w := createWindow("Test").(*window)
	e := widget.NewEntry()
	u, _ := url.Parse("https://testing.fyne")
	h := widget.NewHyperlink("Testing", u)
	b := widget.NewButton("Test", nil)

	w.SetContent(container.NewVBox(e, h, b))

	w.mouseMoved(w.viewport, 10, float64(e.Position().Y+10))
	textCursor := desktop.TextCursor

	t.Logf("Content=%+v", w.Canvas().Content())
	if w.cursor != textCursor {
		t.Logf("cursor wrong %+v, sleeping", w.cursor)
		for i := 0; i < 20; i++ {
			t.Logf("%d Content=%+v", i*5, w.Canvas().Content())
			time.Sleep(5 * time.Millisecond)
		}
	}

	assert.Equal(t, textCursor, w.cursor)`

I ran the test till I had 10 failures, these are the timings
`20
5
20
5
15
20
0
20
15
15`
This gives an avg of 13.5ms.

**Comment by adrianre12** [comment]
Following the code for `w.SetContent()` I have found that it calls this
`func (w *window) RescaleContext() {
	runOnMain(func() {
		w.rescaleOnMain()
	})
}`

I suspect this is the culprit as it does not wait for the rescaleOnMain() to finish.

**Comment by adrianre12** [comment]
I have looked at another test that failed
```
=== RUN   TestGlCanvas_MinSizeShrinkTriggersLayout
    canvas_test.go:317:
                Error Trace:    canvas_test.go:317
                Error:          Not equal:
                                expected: fyne.Size{Width:212, Height:112}
                                actual  : fyne.Size{Width:9, Height:9}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,4 +1,4 @@
                                 (fyne.Size) {
                                - Width: (float32) 212,
                                - Height: (float32) 112
                                + Width: (float32) 9,
                                + Height: (float32) 9
                                 }
                Test:           TestGlCanvas_MinSizeShrinkTriggersLayout
--- FAIL: TestGlCanvas_MinSizeShrinkTriggersLayout (0.38s)
```
This too is showing the wrong Canvas() size and in the test it also does a `SetContent()` followed by an assert.
```
	w.SetContent(content)

	oldCanvasSize := fyne.NewSize(200+3*theme.Padding(), 100+3*theme.Padding())
	assert.Equal(t, oldCanvasSize, c.Size())
```

**Comment by adrianre12** [comment]
I have a hypothisys.
In the first examples log I can see that when the test fails `shouldExpand=true` then 1s later `shouldExpand=false` the only place that `shouldExpand` is read is in loop.go `runGL()` as part of the eventTick.
I think despite all the other parts of `SetContent()` being run synchronously the final resize of the canvas is being handled by the eventTick. Unfortunatly so far I have not found where/how this is happening.

**Comment by adrianre12** [comment]
I think the best clue we have is in the second test when it fails and we get `fyne.Size{Width:9, Height:9}`
Could this be because the contents is reporting a size of 1x1 and then when the padding is added we get 9x9? But how is it failing and how is it getting set later?

**Comment by adrianre12** [comment]
Latest suspect is the events in the draw thread firing while the SetContent() is running. I need to test more but it looks like something in the events is both causing the failure and when it runs again fixes the Size().

**Comment by adrianre12** [comment]
I did not fancy trying to explain what I found in markdown, I have created a [google doc](https://docs.google.com/document/d/1RNy5bzPRjWdzB4BcG_OiQTCQEM6FIlJE2DNRNpLb75k/edit?usp=sharing) instead.
TLDR:
The tests fail when the eventTicker fires while the test is running SetContent() Not sure how but the window.resized() event that is called sets the canvas size to 9x9. Then when the eventTicker runs again resized() correctly sets the size to 212x112

**Comment by adrianre12** [comment]
I had a look at resized() the size is getting set to 9x9 because the size is being taken from glfw and applied back to the fyne window.canvas.  
In the eventTick, tryPollEvents() is run before section to expand the glfw window size, so it will always take two poll events to reflect back a size change. Looking at the logs, normally the tryPollEvent runs resized twice. ie glfw has two events. 
If I am understanding how it works correctly then it seems a little odd. When setting content the canvas and windows sizes are calculated. Then these are ignored and calculated again by the poll events. 
I am not sure how to fix this and am wondering if the tests are even valid as they are not checking the full operation.
Putting a sleep in to wait for the poll events seems a bit naff, but I can't see how to ensure that the tests wait for the change to be propagated through the events.
Would it be possible for set content to set the glfw windows size?

**Comment by andydotxyz** [comment]
Do you think that #960 will help?

**Comment by andydotxyz** [comment]
Sorry to bump this but its been 6 months - can it be completed or should I close the PR?

**Comment by adrianre12** [comment]
Hey Andy,
Sorry, have not been using this email for a while.
I think I had reached the limits of what I could do with the CI. Basically, it had got to a point where there needed to be a design change in how the code ran or a complete re-write of the tests. The sync tests can't work with the async threading.
It's up to you re closing it. Go for it if you think the tests are ok.

Adrian.

**Comment by andydotxyz** [comment]
Oh I see, sorry I didn't understand. Given the issue I linked to I don't think we can resize window when SetContent is called as it goes in a contradictory direction.
Does that make sense? I hope we can brainstorm a solution because this is an important improvement to our tests :)

**Comment by andydotxyz** [comment]
Would it be worth considering enabling it for some tests then solving resize issues in a second step?

**Comment by adrianre12** [comment]
I would enable CI for every test you can, even if you can't do all the platforms.  The reason I started looking at CI, if you remember was that the windows build was broken, but the CI tests were passing. 
I don't know how you would turn just one test on or off, but I'm sure you do.
Re needing to wait for the double poll, I was thinking of a dirty flag that is only cleared on the second poll path, but that depends if you want flags just for testing in the production code.

**Comment by andydotxyz** [comment]
Is it possible to rebase on latest develop so we can better evaluate the current status?

**Comment by adrianre12** [comment]
Is that what you mean?

**Comment by andydotxyz** [comment]
Perfect, thanks.
I guess we have to merge this before the added actions are available?

**Comment by andydotxyz** [comment]
Oh no, we let this slide. It is on the list of priorities to get all tests activated on CI in the coming weeks, so we can ensure quality stays high.

I wonder if there is a way we can pick the changes out onto modern develop branch? I don't know if @adrianre12 is able to help us pick it back up, or if someone else can pick the relevant changes into a new branch?

**Comment by adrianre12** [comment]
I had a look to remind myself of what it is all about. This part is fairly simple and should work if merged assuming the target OS's have not changed too much.
Looking back at the comments the  Windows tests were passing, and the linux tests were randomly failing due to how Fyne works with its multiple async threads. If I remember right it needed some deep rework which was beyond my knowledge of Fyne. Mac, I know nothing about them and it would need someone with a mac to work on it.
I tore down my dev env ages ago so don't have anywhere to work on it.

**Comment by Jacalz** [comment]
@adrianre12 FYI, I think you accidentally merged master into the branch and not develop.

**Comment by andydotxyz** [comment]
I think this has probably been replaced by the recent #3989

**Comment by adrianre12** [comment]
Yes, they do sound similar.
